kernel/src/gdt.rs

```diff
@@ -52,6 +54,51 @@ struct Selectors {
     tss: SegmentSelector,
 }
 
+#[repr(C)]
+pub struct Tss {
+    pub tss: RefCell<TaskStateSegment>,
+    iomap: [u8; 8193],
+}
+
+impl Tss {
+    pub fn new(tss: TaskStateSegment) -> Self {
+        let mut tss = Tss {
+            tss: RefCell::new(tss),
+            iomap: [0xff; 8193],
+        };
+
+        // Absolute values don't matter, only the difference
+        let tss_addr = tss.tss.as_ptr() as usize;
+        let iomap_addr = (&tss.iomap) as *const _ as usize;
+        let iomap_base = (iomap_addr - tss_addr) as u16;
+
+        tss.tss.get_mut().iomap_base = iomap_base;
```

Comment:
I don't think that this approach works. The `iomap_addr` points to your local stack frame, which is no longer valid after the function returns.

Reply:
It does work, as the iomap base is relative to the TSS start address's. This is why it is iomap_addr - tss_addr. Otherwise yes, it wouldn't work.

Reply:
Ah sorry, I confused `RefCell` and `Rc` and thought that the tss lived on the heap.

Reply:
Oh, I may have forgot to add the size of the TSS. Not sure if that's needed... In testing it has worked though.

Reply:
Wait, I don't, nvm.